File: src/app/api/grid.py
"""
Grid Analysis API for Microgrid Optimization.

This module provides REST endpoints for:
- Grid state analysis (voltage, losses, power quality)
- Zone-level aggregated metrics
- Optimization data export for external algorithms
- Battery dispatch commands

Note: P2P trading matching is handled by the API Gateway and Blockchain.
This API focuses on grid physics and optimization support.
"""

from fastapi import APIRouter, Request, HTTPException, Query
from typing import Dict, Any, Optional, List
from pydantic import BaseModel, Field
import logging

logger = logging.getLogger(__name__)

router = APIRouter()


# =============================================================================
# Pydantic Models for Grid Analysis
# =============================================================================

class GridStateResponse(BaseModel):
    """Response model for grid state at a specific meter."""
    meter_id: str
    voltage_pu: float
    voltage_v: float
    frequency_hz: float
    power_factor: float
    thd_voltage: float
    thd_current: float
    is_on_peak: bool
    temperature_c: float


class ZoneStateResponse(BaseModel):
    """Response model for zone-level aggregated state."""
    zone_id: int
    avg_voltage_pu: float
    min_voltage_pu: float
    max_voltage_pu: float
    total_load_kw: float
    total_generation_kw: float
    net_power_kw: float
    meter_count: int
    has_voltage_violation: bool
    has_overload: bool = False


class BatteryDispatchRequest(BaseModel):
    """Request model for battery dispatch command."""
    meter_id: str = Field(..., description="Target meter with battery")
    power_kw: float = Field(..., description="Power in kW. Positive = discharge, Negative = charge")


class BatteryDispatchResponse(BaseModel):
    """Response model for battery dispatch result."""
    success: bool
    meter_id: str
    power_kw: float
    new_battery_level: Optional[float]
    message: str

@router.get("/thailand/data")
async def get_thailand_data(request: Request):
    """Get static structure of Thailand grid (Transformers/Zones and Meters)"""
    engine = getattr(request.app.state, "engine", None)
    if not engine:
         return {"error": "Simulator not initialized"}
    
    # Reuse /api/zones logic but exposed specifically for this demo to ensure clarity
    # Getting zone summary from zoning service which is now initialized with Thailand data
    try:
        zone_summary = engine.zoning_service.get_zone_summary()
        zones = {}
        for zone_id, info in zone_summary.items():
            zones[zone_id] = {
                "zone_id": int(info.zone_id),
                "centroid_lat": float(info.centroid_lat),
                "centroid_lon": float(info.centroid_lon),
                "meter_count": int(info.meter_count),
                "transformer_name": f"TR-{info.zone_id}", # Custom naming
            }
            
        meters = []
        for meter in engine.meters:
            if meter.latitude is not None and meter.longitude is not None:
                zid = getattr(meter, "grid_zone_id", None)
                meters.append({
                    "meter_id": str(meter.meter_id),
                    "latitude": float(meter.latitude),
                    "longitude": float(meter.longitude),
                    "zone_id": int(zid) if zid is not None else None,
                    "meter_type": str(meter.config.get("meter_type", "Unknown")),
                    "contract_capacity": meter.config.get("contract_capacity_kw", 0),
                    "building_area": meter.config.get("building_area_sqm", 0)
                })
        
        return {
            "region": "Phaya Thai, Bangkok",
            "stats": {
                "total_meters": len(meters),
                "total_transformers": len(zones)
            },
            "zones": zones,
            "meters": meters
        }
    except Exception as e:
        import traceback
        import logging
        logging.error(f"Error in get_thailand_data: {e}")
        return {"error": str(e), "traceback": traceback.format_exc()}

@router.get("/zones")
async def get_zones(request: Request):
    """Get K-Means zone data including centroids and meter assignments"""
    engine = getattr(request.app.state, "engine", None)
    if not engine:
        return {"error": "Simulator not initialized"}

    try:
        # Get zone summary from zoning service
        # Check if zoning_service is available
        if not hasattr(engine, "zoning_service"):
             return {"error": "Zoning service not available on engine"}

        zone_summary = engine.zoning_service.get_zone_summary()
        
        # Build zones dict with explicit type casting for JSON serialization
        zones = {}
        for zone_id, info in zone_summary.items():
            zones[zone_id] = {
                "zone_id": int(info.zone_id),
                "centroid_lat": float(info.centroid_lat),
                "centroid_lon": float(info.centroid_lon),
                "meter_count": int(info.meter_count),
                "transformer_name": str(info.transformer_name),
            }
        
        # Build meters list with zone assignments
        meters = []
        for meter in engine.meters:
            if meter.latitude is not None and meter.longitude is not None:
                zid = getattr(meter, "grid_zone_id", None)
                meters.append({
                    "meter_id": str(meter.meter_id),
                    "latitude": float(meter.latitude),
                    "longitude": float(meter.longitude),
                    "zone_id": int(zid) if zid is not None else None,
                    "meter_type": str(meter.config.get("meter_type", "Unknown")),
                })
        
        return {
            "zones": zones,
            "meters": meters,
            "wheeling_charges": engine.zoning_service.get_wheeling_charge_matrix(),
            "loss_factors": engine.zoning_service.get_loss_factor_matrix(),
        }
    except Exception as e:
        import traceback
        import logging
        logging.error(f"Error in get_zones: {e}")
        return {"error": str(e), "traceback": traceback.format_exc()}


@router.get("/grid/status")
async def get_grid_status(request: Request):
    """Get aggregate grid status (Generation, Consumption, Balance)"""
    engine = getattr(request.app.state, "engine", None)
    if not engine:
        return {"error": "Simulator not initialized"}

    try:
        total_generation = 0.0
        total_consumption = 0.0
        active_meters = 0
        
        for meter in engine.meters:
            # Check connection status
            if getattr(meter, "is_connected", False):
                active_meters += 1
                
                # Get latest reading if available
                if hasattr(meter, "last_reading") and meter.last_reading:
                    total_generation += meter.last_reading.energy_generated
                    total_consumption += meter.last_reading.energy_consumed
        
        net_balance = total_generation - total_consumption
        # Simple CO2 calculation (approximate 0.5 kg/kWh for grid offset)
        co2_saved_kg = total_generation * 0.5 

        from datetime import datetime, timezone
        
        return {
            "total_generation": round(total_generation, 4),
            "total_consumption": round(total_consumption, 4),
            "net_balance": round(net_balance, 4),
            "active_meters": active_meters,
            "co2_saved_kg": round(co2_saved_kg, 4),
            "timestamp": datetime.now(timezone.utc).isoformat()
        }
    except Exception as e:
        import traceback
        import logging
        logging.error(f"Error in get_grid_status: {e}")
        return {"error": str(e), "traceback": traceback.format_exc()}


# =============================================================================
# New Grid Analysis Endpoints for Microgrid Optimization
# =============================================================================

@router.get("/grid/state/{meter_id}", response_model=GridStateResponse)
async def get_meter_grid_state(meter_id: str, request: Request):
    """
    Get current grid state for a specific meter.
    
    Returns voltage, frequency, power factor, THD, and other physical parameters.
    """
    engine = getattr(request.app.state, "engine", None)
    if not engine:
        raise HTTPException(status_code=503, detail="Simulator not initialized")
    
    meter = next((m for m in engine.meters if m.meter_id == meter_id), None)
    if not meter:
        raise HTTPException(status_code=404, detail=f"Meter {meter_id} not found")
    
    grid_state = engine.get_grid_state(meter_id)
    
    return GridStateResponse(
        meter_id=meter_id,
        voltage_pu=grid_state.voltage_pu,
        voltage_v=grid_state.voltage_pu * 230.0,
        frequency_hz=grid_state.frequency_hz,
        power_factor=grid_state.power_factor,
        thd_voltage=grid_state.thd_voltage,
        thd_current=grid_state.thd_current,
        is_on_peak=grid_state.is_on_peak,
        temperature_c=grid_state.temperature_c
    )


@router.get("/grid/zone/{zone_id}/state", response_model=ZoneStateResponse)
async def get_zone_state(zone_id: int, request: Request):
    """
    Get aggregated state for a microgrid zone.
    
    Returns voltage statistics, total load/generation, and violation flags.
    """
    engine = getattr(request.app.state, "engine", None)
    if not engine:
        raise HTTPException(status_code=503, detail="Simulator not initialized")
    
    zone_state = engine.get_zone_state(zone_id)
    
    if zone_state.meter_count == 0:
        raise HTTPException(status_code=404, detail=f"Zone {zone_id} not found or has no meters")
    
    return ZoneStateResponse(**vars(zone_state))


@router.get("/grid/analysis")
async def get_grid_analysis(request: Request):
    """
    Perform comprehensive grid analysis.
    
    Runs power flow calculation and returns:
    - Total load and generation
    - Technical losses
    - Voltage violations
    - Optimization recommendations
    """
    engine = getattr(request.app.state, "engine", None)
    if not engine:
        raise HTTPException(status_code=503, detail="Simulator not initialized")
    
    try:
        analysis = engine.analyze_grid()
        
        return {
            "timestamp": analysis.timestamp.isoformat() if analysis.timestamp else "",
            "power_flow_converged": analysis.power_flow_converged,
            "total_load_mw": analysis.total_load_mw,
            "total_generation_mw": analysis.total_generation_mw,
            "total_loss_mw": analysis.total_loss_mw,
            "loss_percentage": analysis.loss_percentage,
            "zone_count": len(analysis.zone_states),
            "voltage_violations": analysis.voltage_violations,
            "overloaded_elements": analysis.overloaded_elements,
            "recommendations": analysis.recommendations
        }
    except Exception as e:
        logger.error(f"Grid analysis failed: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/grid/losses")
async def get_loss_analysis(request: Request):
    """
    Get detailed technical loss analysis.
    
    Returns loss breakdown by zone with optimization recommendations.
    """
    engine = getattr(request.app.state, "engine", None)
    if not engine:
        raise HTTPException(status_code=503, detail="Simulator not initialized")
    
    try:
        loss_data = engine.get_loss_analysis()
        return loss_data
    except Exception as e:
        logger.error(f"Loss analysis failed: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/grid/optimization-data")
async def get_optimization_data(request: Request):
    """
    Get data package for external optimization algorithms.
    
    Returns all necessary information for battery dispatch,
    load scheduling, or DER optimization algorithms.
    """
    engine = getattr(request.app.state, "engine", None)
    if not engine:
        raise HTTPException(status_code=503, detail="Simulator not initialized")
    
    try:
        data = engine.get_optimization_data()
        
        # Add summary stats
        data["summary"] = {
            "total_load_mw": sum(m.get("current_load_kw", 0) for m in data.get("meters", [])) / 1000,
            "total_gen_mw": sum(m.get("current_gen_kw", 0) for m in data.get("meters", [])) / 1000,
            "meters_with_battery": len([m for m in data.get("meters", []) if m.get("has_battery")]),
            "meters_with_solar": len([m for m in data.get("meters", []) if m.get("has_solar")]),
            "available_battery_kwh": sum(
                m.get("battery_capacity", 0) * m.get("battery_level", 0) / 100
                for m in data.get("meters", []) if m.get("has_battery")
            )
        }
        
        return data
    except Exception as e:
        logger.error(f"Failed to get optimization data: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/grid/battery/dispatch", response_model=BatteryDispatchResponse)
async def dispatch_battery(request: Request, dispatch: BatteryDispatchRequest):
    """
    Send battery dispatch command to a meter.
    
    Use positive power_kw to discharge (inject to grid),
    negative power_kw to charge (absorb from grid).
    """
    engine = getattr(request.app.state, "engine", None)
    if not engine:
        raise HTTPException(status_code=503, detail="Simulator not initialized")
    
    meter = next((m for m in engine.meters if m.meter_id == dispatch.meter_id), None)
    if not meter:
        raise HTTPException(status_code=404, detail=f"Meter {dispatch.meter_id} not found")
    
    if not meter.config.get("has_battery"):
        raise HTTPException(
            status_code=400, 
            detail=f"Meter {dispatch.meter_id} does not have a battery"
        )
    
    success = engine.apply_battery_dispatch(dispatch.meter_id, dispatch.power_kw)
    
    action = "discharge" if dispatch.power_kw > 0 else "charge"
    
    return BatteryDispatchResponse(
        success=success,
        meter_id=dispatch.meter_id,
        power_kw=dispatch.power_kw,
        new_battery_level=meter.battery_level if success else None,
        message=f"Battery {action} command {'sent' if success else 'failed'}"
    )


@router.get("/grid/events")
async def get_grid_events(
    request: Request,
    event_type: Optional[str] = Query(None, description="Filter by event type"),
    zone_id: Optional[int] = Query(None, description="Filter by zone"),
    limit: int = Query(100, le=1000, description="Maximum events to return")
):
    """
    Get recent grid events (violations, dispatches, milestones).
    """
    engine = getattr(request.app.state, "engine", None)
    if not engine:
        return []
    
    if not hasattr(engine, "ledger"):
        return []
    
    return engine.ledger.get_events(event_type=event_type, zone_id=zone_id, limit=limit)


@router.get("/grid/health")
async def grid_health_check(request: Request):
    """
    Quick grid health check endpoint.
    
    Returns basic health status without full analysis.
    """
    engine = getattr(request.app.state, "engine", None)
    if not engine:
        return {"healthy": False, "error": "Simulator not initialized"}
    
    is_valid = engine.validate_grid_state()
    
    return {
        "healthy": is_valid,
        "meter_count": len(engine.meters),
        "simulation_time": engine.current_sim_time.isoformat() if engine.current_sim_time else None,
        "model_type": engine.model_type
    }

File: rust_api/src/routes/grid.rs
//! Grid analysis API routes

use actix_web::{web, HttpResponse, Scope};
use std::sync::Arc;
use std::collections::HashMap;
use chrono::Utc;
use rand::Rng;

use crate::models::*;
use crate::state::AppState;

/// Configure grid routes
pub fn configure() -> Scope {
    web::scope("/grid")
        .route("/status", web::get().to(get_grid_status))
        .route("/zones", web::get().to(get_zones))
        .route("/state/{meter_id}", web::get().to(get_meter_grid_state))
        .route("/zone/{zone_id}/state", web::get().to(get_zone_state))
        .route("/zones/state", web::get().to(get_all_zone_states))
        .route("/analysis", web::get().to(get_grid_analysis))
        .route("/losses", web::get().to(get_loss_analysis))
        .route("/optimization-data", web::get().to(get_optimization_data))
        .route("/battery/dispatch", web::post().to(dispatch_battery))
        .route("/events", web::get().to(get_grid_events))
        .route("/health", web::get().to(grid_health))
        .route("/debug/stats", web::get().to(debug_stats))
        .route("/thailand/data", web::get().to(get_thailand_data))
}

/// Get aggregate grid status
async fn get_grid_status(state: web::Data<Arc<AppState>>) -> HttpResponse {
    let meters = state.meters.read().unwrap();
    
    let mut total_generation = 0.0;
    let mut total_consumption = 0.0;
    let mut active_meters = 0;
    
    for meter in meters.values() {
        if meter.is_connected {
            active_meters += 1;
            if let Some(reading) = &meter.last_reading {
                total_generation += reading.energy_generated;
                total_consumption += reading.energy_consumed;
            }
        }
    }
    
    let net_balance = total_generation - total_consumption;
    let co2_saved_kg = total_generation * 0.5;
    
    HttpResponse::Ok().json(GridStatusResponse {
        total_generation: round_to(total_generation, 4),
        total_consumption: round_to(total_consumption, 4),
        net_balance: round_to(net_balance, 4),
        active_meters,
        co2_saved_kg: round_to(co2_saved_kg, 4),
        timestamp: Utc::now().to_rfc3339(),
    })
}

/// Get zones with meter assignments
async fn get_zones(state: web::Data<Arc<AppState>>) -> HttpResponse {
    let zones = state.zones.read().unwrap();
    let meters = state.meters.read().unwrap();
    
    let mut zones_map = HashMap::new();
    for (id, zone) in zones.iter() {
        zones_map.insert(id.to_string(), ZoneInfo {
            zone_id: zone.zone_id,
            centroid_lat: zone.centroid_lat,
            centroid_lon: zone.centroid_lon,
            meter_count: zone.meter_ids.len() as i32,
            transformer_name: zone.transformer_name.clone(),
        });
    }
    
    let meter_list: Vec<MeterZoneInfo> = meters.values()
        .filter(|m| m.latitude.is_some() && m.longitude.is_some())
        .map(|m| MeterZoneInfo {
            meter_id: m.meter_id.clone(),
            latitude: m.latitude.unwrap_or(0.0),
            longitude: m.longitude.unwrap_or(0.0),
            zone_id: m.zone_id,
            meter_type: m.meter_type.clone(),
        })
        .collect();
    
    // Build wheeling/loss matrices
    let mut wheeling_charges = HashMap::new();
    let mut loss_factors = HashMap::new();
    
    for from_zone in zones.keys() {
        for to_zone in zones.keys() {
            let key = format!("{}_{}", from_zone, to_zone);
            wheeling_charges.insert(key.clone(), if from_zone == to_zone { 0.01 } else { 0.05 });
            loss_factors.insert(key, if from_zone == to_zone { 0.01 } else { 0.03 });
        }
    }
    
    HttpResponse::Ok().json(ZonesResponse {
        zones: zones_map,
        meters: meter_list,
        wheeling_charges,
        loss_factors,
    })
}

/// Get grid state for a specific meter
async fn get_meter_grid_state(
    state: web::Data<Arc<AppState>>,
    path: web::Path<String>,
) -> HttpResponse {
    let meter_id = path.into_inner();
    let meters = state.meters.read().unwrap();
    
    if let Some(meter) = meters.get(&meter_id) {
        let mut rng = rand::thread_rng();
        let hour = Utc::now().hour() as i32;
        let is_peak = (hour >= 9 && hour <= 14) || (hour >= 18 && hour <= 22);
        
        HttpResponse::Ok().json(GridStateResponse {
            meter_id: meter.meter_id.clone(),
            voltage_pu: round_to(1.0 + rng.gen_range(-0.02..0.02), 4),
            voltage_v: round_to(230.0 + rng.gen_range(-4.0..4.0), 1),
            frequency_hz: round_to(50.0 + rng.gen_range(-0.01..0.01), 3),
            power_factor: round_to(0.95 + rng.gen_range(-0.02..0.02), 3),
            thd_voltage: round_to(rng.gen_range(1.0..3.0), 2),
            thd_current: round_to(rng.gen_range(3.0..8.0), 2),
            is_on_peak: is_peak,
            temperature_c: round_to(25.0 + rng.gen_range(-2.0..5.0), 1),
        })
    } else {
        HttpResponse::NotFound().json(ErrorResponse::new(format!("Meter {} not found", meter_id)))
    }
}

use chrono::Timelike;

/// Get zone state
async fn get_zone_state(
    state: web::Data<Arc<AppState>>,
    path: web::Path<i32>,
) -> HttpResponse {
    let zone_id = path.into_inner();
    let zones = state.zones.read().unwrap();
    let meters = state.meters.read().unwrap();
    
    if let Some(zone) = zones.get(&zone_id) {
        let mut rng = rand::thread_rng();
        
        let mut total_load = 0.0;
        let mut total_gen = 0.0;
        let mut voltages = Vec::new();
        
        for meter_id in &zone.meter_ids {
            if let Some(meter) = meters.get(meter_id) {
                if let Some(reading) = &meter.last_reading {
                    total_load += reading.energy_consumed * 4.0; // Convert to kW
                    total_gen += reading.energy_generated * 4.0;
                    voltages.push(reading.voltage / 230.0);
                }
            }
        }
        
        let avg_v = if voltages.is_empty() { 1.0 } else { voltages.iter().sum::<f64>() / voltages.len() as f64 };
        let min_v = voltages.iter().cloned().fold(1.0, f64::min);
        let max_v = voltages.iter().cloned().fold(1.0, f64::max);
        
        let has_violation = min_v < 0.95 || max_v > 1.05;
        let has_overload = total_load > 100.0; // Arbitrary threshold
        
        let voltage_penalty = (1.0 - avg_v).abs() * 100.0;
        let mut health = (100.0 - voltage_penalty).max(0.0);
        if has_violation { health *= 0.8; }
        if has_overload { health *= 0.7; }
        
        HttpResponse::Ok().json(ZoneStateResponse {
            zone_id,
            avg_voltage_pu: round_to(avg_v, 4),
            min_voltage_pu: round_to(min_v, 4),
            max_voltage_pu: round_to(max_v, 4),
            total_load_kw: round_to(total_load, 2),
            total_generation_kw: round_to(total_gen, 2),
            net_power_kw: round_to(total_gen - total_load, 2),
            meter_count: zone.meter_ids.len() as i32,
            has_voltage_violation: has_violation,
            has_overload,
            health_score: round_to(health, 2),
        })
    } else {
        HttpResponse::NotFound().json(ErrorResponse::new(format!("Zone {} not found", zone_id)))
    }
}

/// Get all zone states
async fn get_all_zone_states(state: web::Data<Arc<AppState>>) -> HttpResponse {
    let zones = state.zones.read().unwrap();
    let meters = state.meters.read().unwrap();
    
    let mut results = Vec::new();
    
    for (zone_id, zone) in zones.iter() {
        let mut total_load = 0.0;
        let mut total_gen = 0.0;
        let mut voltages = Vec::new();
        
        for meter_id in &zone.meter_ids {
            if let Some(meter) = meters.get(meter_id) {
                if let Some(reading) = &meter.last_reading {
                    total_load += reading.energy_consumed * 4.0;
                    total_gen += reading.energy_generated * 4.0;
                    voltages.push(reading.voltage / 230.0);
                }
            }
        }
        
        let avg_v = if voltages.is_empty() { 1.0 } else { voltages.iter().sum::<f64>() / voltages.len() as f64 };
        let has_violation = voltages.iter().any(|&v| v < 0.95 || v > 1.05);
        let has_overload = total_load > 100.0;
        
        let voltage_penalty = (1.0 - avg_v).abs() * 100.0;
        let mut health = (100.0 - voltage_penalty).max(0.0);
        if has_violation { health *= 0.8; }
        if has_overload { health *= 0.7; }
        
        results.push(serde_json::json!({
            "zone_id": zone_id,
            "health_score": round_to(health, 2),
            "has_overload": has_overload,
            "transformer_name": zone.transformer_name,
            "avg_voltage_pu": round_to(avg_v, 4),
            "net_power_kw": round_to(total_gen - total_load, 2)
        }));
    }
    
    HttpResponse::Ok().json(results)
}

/// Get grid analysis
async fn get_grid_analysis(state: web::Data<Arc<AppState>>) -> HttpResponse {
    let meters = state.meters.read().unwrap();
    let zones = state.zones.read().unwrap();
    
    let mut total_load = 0.0;
    let mut total_gen = 0.0;
    
    for meter in meters.values() {
        if let Some(reading) = &meter.last_reading {
            total_load += reading.energy_consumed * 4.0 / 1000.0; // MW
            total_gen += reading.energy_generated * 4.0 / 1000.0;
        }
    }
    
    let total_loss = total_load * 0.02; // 2% loss estimate
    let loss_pct = if total_load > 0.0 { total_loss / total_load * 100.0 } else { 0.0 };
    
    HttpResponse::Ok().json(GridAnalysisResponse {
        timestamp: Utc::now().to_rfc3339(),
        power_flow_converged: true,
        total_load_mw: round_to(total_load, 4),
        total_generation_mw: round_to(total_gen, 4),
        total_loss_mw: round_to(total_loss, 4),
        loss_percentage: round_to(loss_pct, 2),
        zone_count: zones.len() as i32,
        voltage_violations: Vec::new(),
        overloaded_elements: Vec::new(),
        recommendations: vec!["System operating normally".to_string()],
    })
}

/// Get loss analysis
async fn get_loss_analysis(state: web::Data<Arc<AppState>>) -> HttpResponse {
    let zones = state.zones.read().unwrap();
    
    let mut zone_losses = HashMap::new();
    for (zone_id, _) in zones.iter() {
        zone_losses.insert(zone_id.to_string(), serde_json::json!({
            "zone_id": zone_id,
            "loss_mw": 0.01,
            "loss_percent": 2.0,
            "recommendations": []
        }));
    }
    
    HttpResponse::Ok().json(serde_json::json!({
        "total_loss_mw": 0.03,
        "total_loss_percent": 2.0,
        "zone_losses": zone_losses
    }))
}

/// Get optimization data
async fn get_optimization_data(state: web::Data<Arc<AppState>>) -> HttpResponse {
    let meters = state.meters.read().unwrap();
    
    let mut meter_data = Vec::new();
    let mut total_load = 0.0;
    let mut total_gen = 0.0;
    let mut meters_with_battery = 0;
    let mut meters_with_solar = 0;
    let mut available_battery = 0.0;
    
    for meter in meters.values() {
        let current_load = meter.last_reading.as_ref().map(|r| r.energy_consumed * 4.0).unwrap_or(0.0);
        let current_gen = meter.last_reading.as_ref().map(|r| r.energy_generated * 4.0).unwrap_or(0.0);
        
        total_load += current_load;
        total_gen += current_gen;
        
        let has_battery = meter.battery_capacity > 0.0;
        let has_solar = meter.solar_capacity > 0.0;
        
        if has_battery {
            meters_with_battery += 1;
            available_battery += meter.battery_capacity * meter.battery_level / 100.0;
        }
        if has_solar {
            meters_with_solar += 1;
        }
        
        meter_data.push(serde_json::json!({
            "meter_id": meter.meter_id,
            "zone_id": meter.zone_id,
            "current_load_kw": round_to(current_load, 2),
            "current_gen_kw": round_to(current_gen, 2),
            "has_battery": has_battery,
            "has_solar": has_solar,
            "battery_capacity": meter.battery_capacity,
            "battery_level": meter.battery_level,
            "solar_capacity": meter.solar_capacity
        }));
    }
    
    HttpResponse::Ok().json(serde_json::json!({
        "meters": meter_data,
        "summary": {
            "total_load_mw": round_to(total_load / 1000.0, 4),
            "total_gen_mw": round_to(total_gen / 1000.0, 4),
            "meters_with_battery": meters_with_battery,
            "meters_with_solar": meters_with_solar,
            "available_battery_kwh": round_to(available_battery, 2)
        }
    }))
}

/// Dispatch battery command
async fn dispatch_battery(
    state: web::Data<Arc<AppState>>,
    req: web::Json<BatteryDispatchRequest>,
) -> HttpResponse {
    let mut meters = state.meters.write().unwrap();
    
    if let Some(meter) = meters.get_mut(&req.meter_id) {
        if meter.battery_capacity <= 0.0 {
            return HttpResponse::BadRequest().json(ErrorResponse::new(
                format!("Meter {} does not have a battery", req.meter_id)
            ));
        }
        
        // Apply dispatch (simplified)
        let power_pct = req.power_kw / meter.battery_capacity * 100.0;
        if req.power_kw > 0.0 {
            // Discharge
            meter.battery_level = (meter.battery_level - power_pct.abs()).max(0.0);
        } else {
            // Charge
            meter.battery_level = (meter.battery_level + power_pct.abs()).min(100.0);
        }
        
        let action = if req.power_kw > 0.0 { "discharge" } else { "charge" };
        
        HttpResponse::Ok().json(BatteryDispatchResponse {
            success: true,
            meter_id: req.meter_id.clone(),
            power_kw: req.power_kw,
            new_battery_level: Some(meter.battery_level),
            message: format!("Battery {} command sent", action),
        })
    } else {
        HttpResponse::NotFound().json(ErrorResponse::new(format!("Meter {} not found", req.meter_id)))
    }
}

/// Get grid events
async fn get_grid_events(_state: web::Data<Arc<AppState>>) -> HttpResponse {
    // Return empty events for now
    HttpResponse::Ok().json(Vec::<serde_json::Value>::new())
}

/// Grid health check
async fn grid_health(state: web::Data<Arc<AppState>>) -> HttpResponse {
    let meters = state.meters.read().unwrap();
    let is_running = *state.is_running.read().unwrap();
    let sim_time = state.current_sim_time.read().unwrap();
    
    HttpResponse::Ok().json(GridHealthResponse {
        healthy: true,
        meter_count: meters.len() as i32,
        simulation_time: sim_time.map(|t| t.to_rfc3339()),
        model_type: "rust_simulation".to_string(),
    })
}

/// Debug stats
async fn debug_stats(state: web::Data<Arc<AppState>>) -> HttpResponse {
    let meters = state.meters.read().unwrap();
    let zones = state.zones.read().unwrap();
    
    HttpResponse::Ok().json(serde_json::json!({
        "meter_count": meters.len(),
        "zone_summary_count": zones.len(),
        "zones_keys": zones.keys().collect::<Vec<_>>(),
        "has_zoning_service": true,
        "model_type": "rust_simulation"
    }))
}

/// Get Thailand data (for demo)
async fn get_thailand_data(state: web::Data<Arc<AppState>>) -> HttpResponse {
    let zones = state.zones.read().unwrap();
    let meters = state.meters.read().unwrap();
    
    let mut zones_map = HashMap::new();
    for (id, zone) in zones.iter() {
        zones_map.insert(id.to_string(), serde_json::json!({
            "zone_id": zone.zone_id,
            "centroid_lat": zone.centroid_lat,
            "centroid_lon": zone.centroid_lon,
            "meter_count": zone.meter_ids.len(),
            "transformer_name": format!("TR-{}", zone.zone_id)
        }));
    }
    
    let meter_list: Vec<serde_json::Value> = meters.values()
        .filter(|m| m.latitude.is_some() && m.longitude.is_some())
        .map(|m| serde_json::json!({
            "meter_id": m.meter_id,
            "latitude": m.latitude,
            "longitude": m.longitude,
            "zone_id": m.zone_id,
            "meter_type": m.meter_type,
            "contract_capacity": 15,
            "building_area": 500
        }))
        .collect();
    
    HttpResponse::Ok().json(serde_json::json!({
        "region": "Phaya Thai, Bangkok",
        "stats": {
            "total_meters": meter_list.len(),
            "total_transformers": zones.len()
        },
        "zones": zones_map,
        "meters": meter_list
    }))
}

fn round_to(value: f64, decimals: i32) -> f64 {
    let factor = 10_f64.powi(decimals);
    (value * factor).round() / factor
}
